crates/sage/src/scoring.rs

```diff
@@ -373,15 +376,31 @@ impl<'db> Scorer<'db> {
         }
     }
 
+    fn notched_initial_hits(&self, query: &ProcessedSpectrum) -> (InitialHits, Vec<usize>) {
+        let mut cumsum = 0;
+        let mut hits = InitialHits::default();
+        // Match lengths is pre cumulative sum of the number of hits for each precursor
+        let mut cum_match_lengths = Vec::with_capacity(query.precursors.len());
+
+        for precursor in &query.precursors {
```

Comment:
Is there an alternate approach to this loop that might be simpler?

I am thinking along the lines of:
- Run preliminary search for each precursor
- Build features for each precursor
- Combine all features across precursor, sorting by hyperscore
- Keep `report_psms` features

I think this should approximate what you have here, but with some reduced maintenance burden. Is there something this approach would substantially miss that is implemented here?